src/main/java/de/tum/in/www1/artemis/service/hestia/TestwiseCoverageService.java

```diff
@@ -0,0 +1,291 @@
+        }
+        catch (InterruptedException | GitAPIException e) {
+            log.error("Exception while generating testwise coverage report", e);
+            throw new InternalServerErrorException("Error while generating testwise coverage report: " + e.getMessage());
+        }
+    }
+
+    /**
+     * Calculates the aggregated covered line ratio for all file reports.
+     * @param lineCountByFileName the general line count by file name
+     * @param coveredLineCountByFileName the covered line count by file name
+     * @return the covered line ratio for all files
+     */
+    private double calculateAggregatedLineCoverage(Map<String, Integer> lineCountByFileName, Map<String, Integer> coveredLineCountByFileName) {
+        var aggregatedLineCount = lineCountByFileName.values().stream().mapToInt(Integer::intValue).sum();
+        if (aggregatedLineCount == 0) {
+            return 0;
+        }
+        var aggregatedCoveredLineCount = coveredLineCountByFileName.values().stream().mapToInt(Integer::intValue).sum();
+
+        return aggregatedCoveredLineCount / (double) aggregatedLineCount;
+    }
+
+    /**
+     * Calculate the unique covered line count for all file reports and save this value to the database for all
+     * individual file reports. CoverageFileReports can contain multiple TestwiseCoverageReportEntries referencing
+     * the same lines, but referencing a different test case. This mapping is still required, but simple summing may
+     * count the same covered lines multiple times.
+     * @param report the report for which the line counts of its file reports should be caluclated and saved
+     * @return the number of covered lines by file path
+     */
+    private Map<String, Integer> calculateAndSaveUniqueLineCountsByFilePath(CoverageReport report) {
+        var coveredLinesByFilePath = new HashMap<String, Integer>();
+        report.getFileReports().forEach(fileReport -> {
+            var lineSet = new HashSet<Integer>();
+            fileReport.getTestwiseCoverageEntries()
+                    .forEach(entry -> lineSet.addAll(IntStream.rangeClosed(entry.getStartLine(), entry.getStartLine() + entry.getLineCount() - 1).boxed().toList()));
+            fileReport.setCoveredLineCount(lineSet.size());
+            coverageFileReportRepository.save(fileReport);
+            coveredLinesByFilePath.put(fileReport.getFilePath(), lineSet.size());
+        });
+        return coveredLinesByFilePath;
+    }
+
+    /**
+     * Returns the testwise coverage report for the latest solution submission for a programming exercise without the file reports
+     * @param programmingExercise the exercise for which the latest coverage report should be retrieved
+     * @return the testwise coverage report for the latest solution submission without the file reports
+     */
+    public CoverageReport getCoverageReportForLatestSolutionSubmissionFromProgrammingExercise(ProgrammingExercise programmingExercise) {
```

Comment:
The called repository method does not return an Optional<CoverageReport>, but a List<CoverageReport>.
However, instead of returning null, an empty Optional would be better. What do you think about that?

Reply:
Ah, whoopsie. I only saw "isEmpty" and thought of an Optional :D

Yes, get latest already indicates that it's just one. Optional would make the most sense here.